clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -302,12 +302,22 @@ void runOnce() {
                     transactionManager.transitionToFatalError(
                         new KafkaException("The client hasn't received acknowledgment for " +
                             "some previously sent messages and can no longer retry them. It isn't safe to continue."));
-                } else if (transactionManager.hasInFlightTransactionalRequest() || maybeSendTransactionalRequest()) {
+                } else if (transactionManager.hasInFlightTransactionalRequest()) {
                     // as long as there are outstanding transactional requests, we simply wait for them to return
                     client.poll(retryBackoffMs, time.milliseconds());
                     return;
                 }
 
+                maybeBeginFlush();
+                TransactionManager.TxnRequestHandler nextRequestHandler = transactionManager.nextRequestHandler(accumulator.hasIncomplete());
+                if (nextRequestHandler != null) {
+                    if (maybeSendTransactionalRequest(nextRequestHandler)) {
+                        client.poll(retryBackoffMs, time.milliseconds());
```

Comment:
This logic is a bit awkward. I guess it's because `client.poll` is disconnected from the send logic. I wonder if it would be reasonable to create a method like `maybePollTransactionalRequest` which handles both the inflight check, sending the next request, and calling `client.poll`.